### ISSUE TITLE
Remove core feature

### ISSUE THREAD

**Issue by vadixidav** [arrayfire/arrayfire-rust#189]
The `core` feature in `Cargo.toml` should be removed. If you try to build without it then the build fails:

```rust
error[E0432]: unresolved import `crate::util`
   --> .../src/lib.rs:140:16
    |
140 | pub use crate::util::{get_size, HasAfEnum, ImplicitPromote};
    |                ^^^^ maybe a missing `extern crate util;`?
```

Unless this is just a bug (which would need to be fixed anyways), I would recommend removing this feature. All features of a cargo package should be additive and compatible with each other without breaking the build.

**Comment by 9prady9** [comment]
I am not sure I understand your concern. I added these features to enable compiling sections/modules of code without having to change entire code base when repo-wide changes are being implemented.

Also, as the name suggests core is mandatory feature that is needed to build any other feature. That is the idea. 

Why do you think it needs to be removed?

**Comment by 9prady9** [comment]
Ah, so if it is not included in any feature, it would be considered mandatory for any other feature being compiled? Is that what you mean?

**Comment by vadixidav** [comment]
Perhaps. I mean that core should just be included always and should not have a feature. The reason I say this is because if you omit all features then the crate does not build. If I cannot remove the core module then it shouldn't be an option. It should just always be on.

**Comment by 9prady9** [comment]
I will look into what can be done. Core is useful in certain Dev situations like the one I explained earlier. I will check cargo docs if we can retain but it but make that feature mandatory

**Comment by vadixidav** [comment]
@9prady9 Basically my suggestion is to remove the core feature and all `cfg` statements for it in the code. Those pieces of code are necessary for `arrayfire` to build, therefore it is not useful as a feature since it cannot be turned on or off. Since it can't be turned on or off, it isn't a feature, but rather something required for `arrayfire`, thus it should not be possible to turn off by disabling the feature. It is expected that crates build even when all features are disabled.

**Comment by 9prady9** [comment]
I understood your point. May I see the the cargo command that lead you to the error? Is it with `--no-default-features`

**Comment by vadixidav** [comment]
Actually, it is downstream:

```toml
[dependencies.arrayfire]
git = "https://github.com/arrayfire/arrayfire-rust.git"
rev = "5c5013583d0c518860d2a313204ff80dcaeba8ec"
default-features = false
```

**Comment by 9prady9** [comment]
That's fine, just wanted to how did it happen. So you are using cargo equivalent of command line option for no default feature set.

**Comment by 9prady9** [comment]
Interesting, where are specifying the required features then? In the command line?

**Comment by vadixidav** [comment]
I am specifying them in my `Cargo.toml`. I have to include the `core` feature for it to build:

```toml
[dependencies.arrayfire]
git = "https://github.com/arrayfire/arrayfire-rust.git"
rev = "5c5013583d0c518860d2a313204ff80dcaeba8ec"
default-features = false
features = ["core"]
```

Naturally, I will be adding more features to the features list, but I thought it was weird that `arrayfire` failed to compile with all the features off, and it took me a while to figure out why. It is definitely unexpected and the error is not helpful in determining the cause.

**Comment by vadixidav** [comment]
I can submit a PR to remove the feature if you are okay with the removal.

**Comment by 9prady9** [comment]
I was just wondering if there is way to specify mandatory feature in cargo. Sure send the PR. You just need to edit couple of lines in `src/lib.rs`.

I am still not sure what should happen when we compile empty crate in rust which is essentially what you were earlier.

**Comment by 9prady9** [comment]
Fixed in #190
